### PR TITLE
ImportC: automatically #include importc.h

### PR DESCRIPTION
**Pull request by WalterBright**
Each preprocessor does it differently.

### CODE REVIEW

File: src/dmd/dmodule.d [hunk 1]
```diff
@@ -677,7 +677,33 @@ extern (C++) final class Module : Package
...
+             * It should be in the same place as "object.d"
+             */
+            const(char)* importc_h;
+
+            foreach (entry; (global.path ? (*global.path)[] : null))
+            {
+                auto f = FileName.combine(entry, "importc.h");
+                if (FileName.exists(f) == 1)
+                {
+                     importc_h = f;
```
Comment (ljmf00) [inline src/dmd/dmodule.d:690]: Where did you clean this? It is going to leak.

**Review by ljmf00 — DISMISSED** [review]

File: src/dmd/dmodule.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/dmodule.d:690]: `FileName.combine()` may or may not allocate memory, so it cannot be free'd.

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -677,7 +677,33 @@ extern (C++) final class Module : Package
...
-            filename = global.preprocess(srcfile, ifile);  // run C preprocessor
+            /* Look for "importc.h" by searching along import path.
+             * It should be in the same place as "object.d"
+             */
+            const(char)* importc_h;
+
+            foreach (entry; (global.path ? (*global.path)[] : null))
+            {
+                auto f = FileName.combine(entry, "importc.h");
+                if (FileName.exists(f) == 1)
```
Comment (dkorpel) [inline src/dmd/dmodule.d:688]: I thought this would be redundant, but it turns out `exists` returns a magic number

**Review by dkorpel — APPROVED** [review]

File: src/dmd/dmodule.d (hunk 1 above)
Comment (ljmf00) [reply src/dmd/dmodule.d:690]: So you mean we may leak? What `FileName.free()` does? It seems bad design if we leak resources.

File: src/dmd/dmodule.d [hunk 3]
```diff
@@ -677,7 +677,33 @@ extern (C++) final class Module : Package
...
+
+            if (importc_h)
+            {
+                if (global.params.verbose)
+                    message("include   %s", importc_h);
+            }
+            else
+            {
+                error("cannot find \"importc.h\" along import path");
+                fatal();
```
Comment (ibuclaw) [inline src/dmd/dmodule.d:704]: :-(

**Comment by ibuclaw** [comment]
Why is the burden placed on the callee of preprocess in order to find this header?

**Pull request merged**